spec/scenarios/subscriptions/free_trial_billing_spec.rb

```diff
@@ -0,0 +1,217 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe 'Free Trial Billing Subscriptions Scenario', :scenarios, type: :request do
+  let(:organization) { create(:organization, webhook_url: nil) }
+
+  let(:timezone) { 'UTC' }
+  let(:customer) { create(:customer, organization:, timezone:) }
+
+  let(:plan) do
+    create(
+      :plan,
+      organization:,
+      trial_period:,
+      amount_cents: 5_000_000,
+      pay_in_advance: true,
+    )
+  end
+
+  context 'without free trial' do
+    let(:trial_period) { 0 }
+
+    it 'bills the customer at the beginning of the subscription' do
+      travel_to(Time.zone.parse('2024-03-05T12:12:00')) do
+        create_subscription(
+          {
+            external_customer_id: customer.external_id,
+            external_id: customer.external_id,
+            plan_code: plan.code,
+          },
+        )
+
+        expect(customer.reload.invoices.count).to eq(1)
+        expect(customer.invoices.first.fees.subscription).to exist
+      end
+    end
+  end
+
+  context 'with free trial' do
+    let(:trial_period) { 10 }
+
+    it 'bills the customer at the end of the free trial' do
+      travel_to(Time.zone.parse('2024-03-05T12:12:00')) do
+        create_subscription(
+          {
+            external_customer_id: customer.external_id,
+            external_id: customer.external_id,
+            plan_code: plan.code,
+          },
+        )
+
+        expect(customer.reload.invoices.count).to eq(0)
+      end
+      subscription = customer.subscriptions.sole
+
+      # Ensure nothing happened
+      travel_to(Time.zone.parse('2024-03-10T12:12:00')) do
+        perform_billing
+        expect(customer.reload.invoices.count).to eq(0)
+      end
+
+      # NOTE: The subscription was started at 12:12:00, so the trial period ends exactly at 12:12:00
+      #       This ensure that Subscriptions::FreeTrialBillingService grabs subscriptions that
+      #       ended in the last hour.
+      travel_to(Time.zone.parse('2024-03-15T12:02:00')) do
+        expect(subscription).to be_in_trial_period
+        perform_billing
+        expect(customer.reload.invoices.count).to eq(0)
+      end
+
+      travel_to(Time.zone.parse('2024-03-15T13:02:00')) do
+        expect(subscription).not_to be_in_trial_period
+        perform_billing
+        expect(customer.reload.invoices.count).to eq(1)
+        invoice = customer.reload.invoices.sole
+        expect(invoice.fees.count).to eq(1)
+        expect(invoice.fees.subscription.first.amount_cents).to eq(2_741_935) # (31 - 4 - 10) / 31 * 5_000_000 = 2_741_935
```

Comment:
Layout/LineLength: Line is too long. [122/120] (https://rubystyle.guide#max-line-length)